helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkRoutingDataReader.java

```diff
@@ -22,54 +22,161 @@
+                this);
+      }
+    }
   }
 
-  public Map<String, List<String>> getRoutingData() throws InvalidRoutingDataException {
+  /**
+   * Returns (realm, list of ZK path sharding keys) mappings.
+   * @return
+   * @throws InvalidRoutingDataException
+   */
+  public Map<String, List<String>> getRoutingData()
+      throws InvalidRoutingDataException {
     Map<String, List<String>> routingData = new HashMap<>();
     List<String> children;
     try {
-      children = _zkClient.getChildren(ROUTING_DATA_PATH);
+      children = _zkClient.getChildren(MetadataStoreRoutingConstants.ROUTING_DATA_PATH);
     } catch (ZkNoNodeException e) {
-      throw new InvalidRoutingDataException("Routing data directory ZNode " + ROUTING_DATA_PATH
-          + " does not exist. Routing ZooKeeper address: " + _zkAddress);
+      throw new InvalidRoutingDataException(
+          "Routing data directory ZNode " + MetadataStoreRoutingConstants.ROUTING_DATA_PATH
+              + " does not exist. Routing ZooKeeper address: " + _zkAddress);
     }
     if (children == null || children.isEmpty()) {
       throw new InvalidRoutingDataException(
           "There are no metadata store realms defined. Routing ZooKeeper address: " + _zkAddress);
     }
     for (String child : children) {
-      ZNRecord record = _zkClient.readData(ROUTING_DATA_PATH + "/" + child);
-      List<String> shardingKeys = record.getListField(ZNRECORD_LIST_FIELD_KEY);
+      ZNRecord record =
+          _zkClient.readData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + child);
+      List<String> shardingKeys =
+          record.getListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY);
       if (shardingKeys == null || shardingKeys.isEmpty()) {
-        throw new InvalidRoutingDataException("Realm address ZNode " + ROUTING_DATA_PATH + "/"
-            + child + " does not have a value for key " + ZNRECORD_LIST_FIELD_KEY
-            + ". Routing ZooKeeper address: " + _zkAddress);
+        throw new InvalidRoutingDataException(
+            "Realm address ZNode " + MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + child
+                + " does not have a value for key "
+                + MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY
+                + ". Routing ZooKeeper address: " + _zkAddress);
       }
       routingData.put(child, shardingKeys);
     }
     return routingData;
   }
 
-  public void close() {
+  public synchronized void close() {
+    _zkClient.unsubscribeAll();
     _zkClient.close();
   }
+
+  @Override
+  public synchronized void handleDataChange(String s, Object o)
+      throws Exception {
+    if (_zkClient.isClosed()) {
+      return;
+    }
+    _routingDataListener.updateRoutingData(_namespace);
+  }
+
+  @Override
+  public synchronized void handleDataDeleted(String s)
+      throws Exception {
+    if (_zkClient.isClosed()) {
+      return;
+    }
+    _zkClient.unsubscribeAll();
```

Comment:
Can you share why we need to unsubscribe all here?

Reply:
This is meant to go under handleChildChange. Good catch!